aws/src/main/java/org/apache/iceberg/aws/lakeformation/LakeFormationCredentialsCache.java

```diff
@@ -0,0 +1,151 @@
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg.aws.lakeformation;
+
+import com.github.benmanes.caffeine.cache.LoadingCache;
+import com.github.benmanes.caffeine.cache.Policy;
+import com.github.benmanes.caffeine.cache.stats.CacheStats;
+import java.time.Clock;
+import java.time.Duration;
+import java.util.Map;
+import java.util.concurrent.ConcurrentMap;
+import java.util.concurrent.TimeUnit;
+import java.util.function.Function;
+import javax.annotation.Nullable;
+import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
+import org.checkerframework.checker.index.qual.NonNegative;
+import org.checkerframework.checker.nullness.qual.NonNull;
+
+class LakeFormationCredentialsCache
+    implements LoadingCache<String, LakeFormationTemporaryCredentials> {
+  private final LoadingCache<String, LakeFormationTemporaryCredentials> cache;
+  private final Clock clock;
+  private final Duration expiryLeadTimeForRefresh;
+
+  LakeFormationCredentialsCache(LoadingCache<String, LakeFormationTemporaryCredentials> cache) {
+    this.cache = Preconditions.checkNotNull(cache, "Cache cannot be null");
+    this.clock = Clock.systemUTC();
+    this.expiryLeadTimeForRefresh = Duration.ofMillis(TimeUnit.MINUTES.toMillis(1));
+  }
+
+  boolean needsRefresh(LakeFormationTemporaryCredentials credentials) {
+    if (credentials == null) {
+      return true;
+    }
+
+    if (credentials
+        .sessionCredentialsExpiration()
+        .minus(expiryLeadTimeForRefresh)
+        .isBefore(clock.instant())) {
+      return true;
+    }
+    return false;
+  }
+
+  @Override
+  public LakeFormationTemporaryCredentials get(String key) {
+    return cache.get(key);
+  }
+
+  @Override
+  public Map<String, LakeFormationTemporaryCredentials> getAll(Iterable<? extends String> keys) {
+    return cache.getAll(keys);
+  }
+
+  @Override
+  public void refresh(String key) {
+    cache.refresh(key);
+  }
+
+  @Override
+  public ConcurrentMap<String, LakeFormationTemporaryCredentials> asMap() {
+    return cache.asMap();
+  }
+
+  @Override
+  public void cleanUp() {
+    cache.cleanUp();
+  }
+
+  @Override
+  public @NonNull Policy<String, LakeFormationTemporaryCredentials> policy() {
+    return null;
+  }
+
+  public void refreshIfNecessary(String key) {
```

Comment:
This can be a part of the logic inside `get`? I don't see it called in other places

Reply:
Fixed, thanks!